### PR TITLE
add_rotary

### PR DESCRIPTION
**Pull request by srn-source**
This PR I copy original model GPT2 from huggingface to edit Rotary Encoding

Followed by GPT-NEO-X 
1. I remove origin wpe (original positional encoding)
2. I add function Rotary by Copy from github huggingface (GPT-NEO-X)
3. I change __init__ and forward at  GPT2Model, GPT2Block, GPT2Attention

### CODE REVIEW

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py [hunk 1]
```diff
@@ -28,8 +68,1017 @@ def _attn_wrapper(self, query, key, value, attention_mask=None, head_mask=None):
     return attn_out, None
 
 
+def load_tf_weights_in_gpt2(model, config, gpt2_checkpoint_path):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py:71]: You are copied this whole function from [huggingface gpt2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_gpt2.py). Please import the function from huggingface.

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py [hunk 2]
```diff
@@ -28,8 +68,1017 @@ def _attn_wrapper(self, query, key, value, attention_mask=None, head_mask=None):
...
+            ), f"Pointer shape {pointer.shape} and array shape {array.shape} mismatched"
+        except AssertionError as e:
+            e.args += (pointer.shape, array.shape)
+            raise
+        logger.info(f"Initialize PyTorch weight {name}")
+        pointer.data = torch.from_numpy(array)
+    return model
+
+
+class GPT2PreTrainedModel(PreTrainedModel):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py:125]: How different between this class and [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_gpt2.py)?

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py [hunk 3]
```diff
@@ -28,8 +68,1017 @@ def _attn_wrapper(self, query, key, value, attention_mask=None, head_mask=None):
...
+                    ),
+                )
+
+    def _set_gradient_checkpointing(self, module, value=False):
+        if isinstance(module, GPT2Model):
+            module.gradient_checkpointing = value
+
+
+@dataclass
+class GPT2DoubleHeadsModelOutput(ModelOutput):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py:166]: Where are you use this?

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model_old.py
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/nanoGPT/model_old.py:1]: Please use the better name.

**Review by boat1603 — COMMENTED** [review]
You should refactor `nanoGPT/model.py`. If you directly used some function or class from `transformers`, you should use directly imported.

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model_old.py
Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model_old.py:1]: I have changed to model_without_rotary [Done]

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py (hunk 3 above)
Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py]: Removed [Done]

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py (hunk 2 above)
Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py]: Not different and use directly import instead [Done]

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py [hunk 4]
```diff
@@ -28,8 +61,912 @@ def _attn_wrapper(self, query, key, value, attention_mask=None, head_mask=None):
...
+
+        if use_cache:
+            outputs = (hidden_states,) + outputs
+        else:
+            outputs = (hidden_states,) + outputs[1:]
+
+        return outputs  # hidden_states, present, (attentions, cross_attentions)
+
+
+class RotaryEmbedding(torch.nn.Module):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py:429]: You copied this class from `gpt_neox` in huggingface without modified. You should import from huggingface.

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py [hunk 5]
```diff
@@ -28,8 +61,912 @@ def _attn_wrapper(self, query, key, value, attention_mask=None, head_mask=None):
...
+            freqs = torch.einsum("i,j->ij", t, self.inv_freq)
+            emb = torch.cat((freqs, freqs), dim=-1).to(x.device)
+            self.cos_cached = emb.cos()[None, None, :, :]
+            self.sin_cached = emb.sin()[None, None, :, :]
+        return self.cos_cached[:seq_len, ...].to(x.device), self.sin_cached[
+            :seq_len, ...
+        ].to(x.device)
+
+
+def rotate_half(x):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py:463]: Ditto

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py [hunk 6]
```diff
@@ -28,8 +61,912 @@ def _attn_wrapper(self, query, key, value, attention_mask=None, head_mask=None):
...
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py:470]: Ditto

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model_without_rotary.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import torch.backends.cuda as cuda
+from transformers import AutoConfig
+from transformers.models.gpt2.modeling_gpt2 import GPT2LMHeadModel, GPT2Attention
+
+
+_attn_orig = GPT2Attention._attn
+
+
+# patch GPT2Attention to use flash_sdp, disable it when doing the inference
+def _attn_wrapper(self, query, key, value, attention_mask=None, head_mask=None):
```
[the remark this answers is not in the document]
Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model_without_rotary.py:11]: This is [NOT USE File] just backup if model.py is not work well

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py (hunk 4 above)
Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py]: [Done]

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py (hunk 5 above)
Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py]: [Done]

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py (hunk 6 above)
Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py]: [Done]

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py [hunk 8]
```diff
@@ -28,8 +61,912 @@ def _attn_wrapper(self, query, key, value, attention_mask=None, head_mask=None):
     return attn_out, None
 
 
+class GPT2Attention(nn.Module):
```
[the remark this answers is not in the document]
Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py]: [Done]

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py [hunk 9]
```diff
@@ -28,8 +59,490 @@ def _attn_wrapper(self, query, key, value, attention_mask=None, head_mask=None):
     return attn_out, None
 
 
+class EditGPT2Attention(OriginalGPT2Attention):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py:62]: I think `GPT2AttentionWithRotary` is better name.

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py [hunk 10]
```diff
@@ -28,8 +59,490 @@ def _attn_wrapper(self, query, key, value, attention_mask=None, head_mask=None):
...
+        self.use_rotary = config.use_rotary
+        self.rotary_ndims = int(self.head_dim * self.rotary_pct)
+        if self.use_rotary:
+            self.rotary_emb = RotaryEmbedding(
+                self.rotary_ndims,
+                config.max_position_embeddings,
+                base=self.rotary_emb_base,
+            )
+
+    def forward(
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py:77]: You should add the comment where are you edited from original source code. You can see an example in [Electra repo](https://github.com/OpenThaiGPT/Pretraining-ELECTRA/blob/36acaf52364c4159adfdb2a5e975bd2600f92315/src/electra_pretraining/electra/flash_attn_model.py#L189).

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py (hunk 9 above)
Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py]: DONE

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py (hunk 10 above)
Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py:77]: [DONE] but if you want to make sure that where I edit and add more.... you can search "self.use_rotary" that is the place I'm woking on.

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model_without_rotary.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (boat1603) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model_without_rotary.py:11]: I think `git` is already stored all of version of file. If you have a problem about code in `model.py` you can reverse the old version back.

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py (hunk 9 above)
Comment (boat1603) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py]: You should changed all of class name in your code.

Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model.py]: DONE

File: src/model/openthaigpt_pretraining_model/models/nanoGPT/model_without_rotary.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/models/nanoGPT/model_without_rotary.py]: that's good.. i dont have much experience in git. [DONE removed]

**Review by boat1603 — APPROVED** [review]

File: src/model/openthaigpt_pretraining_model/optimizers/lion/utils.py [hunk 11]
```diff
@@ -206,7 +206,7 @@ def __init__(
             )
         else:
             raise NotImplementedError("only support lion or AdamW")
-        self.model = torch.compile(model)
+        self.model = torch.compile(model)  # type: ignore
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/optimizers/lion/utils.py:209]: If I remembered correctly, you once reported that torch.compile is not working with Flash Attention/Lion.
Could you elaborate why it works now?

Comment (srn-source) [reply src/model/openthaigpt_pretraining_model/optimizers/lion/utils.py:209]: It works correctly. I just told you that i cannot install torch 2 in kaggle meaning that i cannot use torch.compile and flash attention. But on Colab i can install it which cannot use ddp mode to test.

**Pull request merged**